packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -2313,18 +2367,79 @@ Hardhat Network's forking functionality only works with blocks from at least spu
 
     return { maxFeePerGas, maxPriorityFeePerGas };
   }
-}
 
-function getCommonForTracing(networkId: number, blockNumber: number): Common {
-  try {
-    const common = new Common({ chain: networkId });
+  private _selectHardfork(
+    blockNumber: BN,
+    defaultName = this._vm._common.hardfork() as HardforkName
+  ): HardforkName {
+    if (
+      this._forkBlockNumber === undefined ||
+      blockNumber.gte(new BN(this._forkBlockNumber))
+    ) {
+      return defaultName;
+    }
+    if (this._hardforkActivations === undefined) {
+      throw new InternalError(
```

Comment:
I'm pretty sure this error will never be thrown. The reason is that `this._hardforkActivations` can never be undefined because it's initialized like this:

```typescript
let hardforkActivations: HardforkHistoryConfig = new Map();
```

The fix is just changing that line to:

```typescript
    let hardforkActivations: HardforkHistoryConfig | undefined;
```

(I verified this locally.)

Reply:
I took a slightly different approach. I decided to change the check rather than the initialization, in [this commit](https://github.com/nomiclabs/hardhat/pull/1928/commits/6acba23bf668ecc283032382cfac936f354c64b5), because Pato has previously expressed that there should definitely always be a hardfork history given in node construction, because the optionality only exists at the `UserConfig` layer.  What do you think?

Reply:
Yes, that should work too :+1: